### PR TITLE
[flutter_tools] ensure emulator command does not crash with missing avdmanager

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

The avdmanager is not currently verified as a part of the android doctor check. It also may not been in the initial android sdk download we recommend, and only downloaded later on by Android Studio. Update `flutter emulators` to check if it exists before attempting to launch an emulator.

https://github.com/flutter/flutter/issues/57483
https://github.com/flutter/flutter/issues/55971

As a follow up, a check for avdmanager needs to be added to flutter doctor along with download instructions.

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/android_emulator.dart [hunk 1]
```diff
@@ -5,36 +5,131 @@
...
+
   @override
   bool get supportsPlatform => true;
 
   @override
-  bool get canListAnything => androidWorkflow.canListEmulators;
+  bool get canListAnything => _androidWorkflow.canListEmulators;
+
+  @override
+  bool get canLaunchAnything => _androidWorkflow.canListEmulators
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/android_emulator.dart:51]: the `emulators` command for listing emulator images is not the same as the `avdmanager` command for launching/creating them.

File: packages/flutter_tools/lib/src/android/android_emulator.dart [hunk 2]
```diff
@@ -5,36 +5,131 @@
...
+    && _androidSdk.getAvdManagerPath() != null;
 
   @override
-  Future<List<Emulator>> get emulators async => getEmulatorAvds();
+  Future<List<Emulator>> get emulators => _getEmulatorAvds();
+
+  /// Return the list of available emulator AVDs.
+  Future<List<AndroidEmulator>> _getEmulatorAvds() async {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/android/android_emulator.dart:58]: These are the same, just moved into the class

File: packages/flutter_tools/lib/src/emulator.dart [hunk 3]
```diff
@@ -80,6 +101,11 @@ class EmulatorManager {
         name = '${autoName}_${++suffix}';
       }
     }
+    if (!_androidEmulators.canLaunchAnything) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/emulator.dart:104]: This is the check that prevents the crash

File: packages/flutter_tools/test/general.shard/emulator_test.dart [hunk 4]
```diff
@@ -20,106 +20,252 @@ import '../src/common.dart';
 import '../src/context.dart';
 import '../src/mocks.dart';
 
+const FakeEmulator emulator1 = FakeEmulator('Nexus_5', 'Nexus 5', 'Google');
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/emulator_test.dart:23]: I hoisted the logic out of the MockProcessManager in favor of listing out the exact commands with the FakeProcessManager

**Comment by jonahwilliams** [comment]
Note: I verified that this at least prevents a crash on my computer, which despite having a clean doctor does not have avdmanager installed

**Comment by jonahwilliams** [comment]
I was going to refactor and then fix, but when I went to test it crashed due to missing avdmanager

**Review by DanTup — COMMENTED** [review]
LGTM!

**Review by DanTup — APPROVED** [review]

**Pull request merged**